### PR TITLE
984 refactoring separate planning and execution phases in key dataset manipulation services

### PR DESCRIPTION
**Pull request by zaychenko-sergei**
## Description

Closes: #984 

Extracted "planner" and "executor" for compacting, reset, set watermark, push ingest, partially polling ingest.
Renamed long-running "executors" to "agents".
Introduced `MetadataQueryService` to absorb simple queries that do not have to be defined at the level of metadata chian from the interface point of view.

## Checklist before requesting a review

- [x] Unit and integration tests added
  
- [x] Compatibility:
    
  - [x] Network APIs: ✅
    
  - [x] Workspace layout and metadata: ✅
    
  - [x] Configuration: ✅
    
  - [x] Container images: ✅
- [ ] Observability:
  - [x] Tracing / [metrics](https://github.com/kamu-data/kamu-standards/blob/master/metrics_design.md): ✅
  - [ ] Alerts: ✅
- [ ] Documentation:
  - [ ] [Changelog](./CHANGELOG.md): ✅ - TODO
    
  - [ ] [Public documentation](https://github.com/kamu-data/kamu-docs/): ✅ - NO EFFECT
- [ ] Downstream effects:
  - [x] [kamu-node](https://github.com/kamu-data/kamu-node): ✅ - TODO
  - [ ] [kamu-web-ui](https://github.com/kamu-data/kamu-web-ui): ✅ - NO EFFECT
  - [x] [release-queue](https://github.com/orgs/kamu-data/projects/4/views/1) - TODO

### CODE REVIEW

File: src/infra/core/src/services/metadata_query_service_impl.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[component(pub)]
+#[interface(dyn MetadataQueryService)]
+pub struct MetadataQueryServiceImpl {}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl MetadataQueryService for MetadataQueryServiceImpl {
```
Comment (sergiimk) [inline src/infra/core/src/services/metadata_query_service_impl.rs:25]: Would these methods be better suited in `MetadataChainExt` extension trait?

To me a service encapsulates complex interactions between several components, but here we have nothing but a stateless wrapper on top of `MetadataChain`'s visitor interface.

`MetadataChain` is a place where we also validate complex domain rules about what sequences of events are valid an which aren't (e.g. two active push sources with same name) so I don't think an argument that this is some "higher-level logic" works in this case either.

Comment (zaychenko-sergei) [reply src/infra/core/src/services/metadata_query_service_impl.rs:25]: I'm not sure. I have a feeling that for high level queries to dataset's current state, we shouldn't be considering that there is some metadata chain behind it. We just want to know key characteristics of the dataset's current state. The fact that this maps to `MetadataChain` algorithm is an implementation detail. Besides, these questions could be answered with database in future.

Comment (zaychenko-sergei) [reply src/infra/core/src/services/metadata_query_service_impl.rs:25]: `MetadataChainExt` deals with a much lower level of abstraction. This is mostly about visiting the blocks one way or another,  emphasizing explicit knowledge of chain structure.

File: src/infra/ingest-datafusion/src/writer.rs [hunk 2]
```diff
@@ -857,16 +841,16 @@ impl DataWriter for DataWriterDataFusion {
...
-    dataset: Arc<dyn Dataset>,
+    target: ResolvedDataset,
     ctx: SessionContext,
     block_ref: BlockRef,
     metadata_state: Option<DataWriterMetadataState>,
 }
 
 impl DataWriterDataFusionBuilder {
-    pub fn new(dataset: Arc<dyn Dataset>, ctx: SessionContext) -> Self {
+    pub fn new(target: ResolvedDataset, ctx: SessionContext) -> Self {
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/infra/ingest-datafusion/src/writer.rs:851]: Yes, simplified!

**Pull request merged**